store/store_k8s.go

```diff
@@ -0,0 +1,257 @@
+				if msgs := validation.IsDNS1123Subdomain("a" + string(z) + "a"); len(msgs) == 0 {
+					sanitizedString += string(z)
+				} else {
+					sanitizedString += "."
+				}
+			}
+			//After sanitizing, we need the first and last letter to be alphanumeric
+			isAlpha := regexp.MustCompile(`^[A-Za-z0-9]+$`).MatchString
+			first := float64(len(sanitizedString))
+			last := float64(0)
+			for i, s := range sanitizedString {
+				if isAlpha(string(s)) {
+					first = float64(math.Min(first, float64(i)))
+					last = float64(math.Max(last, float64(i)))
+				}
+			}
+			sanitizedString = sanitizedString[int(first) : int(last)+1]
+			logrus.Infof("k8s name is not per RFC 1123 standard sanitized %s to %s", k8sName, sanitizedString)
+		} else {
+			sanitizedString = k8sName
+		}
+	}
+	return sanitizedString
+}
+
+var (
+	// total wait time: 16.25 seconds
+	waitBackoff = wait.Backoff{
+		Duration: waitDuration, // the base duration
+		Factor:   waitFactor,   // Duration is multiplied by factor each iteration
+		Steps:    waitSteps,    // Exit with error after this many steps
+	}
+	etcdErrorsToRetry = []error{rpctypes.ErrLeaderChanged}
+)
+
+type k8sStore struct {
+	cm configmap.ConfigMap
+}
+
+// NewK8sStore returns a Store implementation which uses
+// k8s configmaps to store data.
+func NewK8sStore(clusterID string) (Store, configmap.ConfigMap, error) {
+	k8sStore, cm, err := NewK8sStoreWithParams(
+		configmap.GetName(confgMapPrefix, clusterID),
+		configmap.DefaultK8sLockTimeout,
+		configmap.DefaultK8sLockAttempts * time.Second,
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+	return k8sStore, cm, nil
+}
+
+// NewK8sStoreWithParams returns a Store implementation which uses
+// k8s configmaps to store data. ConfigMap properties can be customized.
+func NewK8sStoreWithParams(
+	name string,
+	lockTryDuration time.Duration,
+	lockTimeout time.Duration,
+) (Store, configmap.ConfigMap, error) {
+	lockAttempts := uint((lockTryDuration / time.Second))
+	cm, err := configmap.New(
+		name,
+		nil,
+		lockTimeout,
+		lockAttempts,
+		0,
+		0,
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+	return &k8sStore{cm}, cm, nil
+}
+
+func (k8s *k8sStore) Lock(owner string) (*StoreLock, error) {
+	if err := k8s.cm.Lock(owner); err != nil {
+		return nil, err
+	}
+	return &StoreLock{owner: owner}, nil
+}
+
+func (k8s *k8sStore) LockWithKey(owner, key string) (*StoreLock, error) {
+	if err := k8s.cm.LockWithKey(owner, key); err != nil {
+		return nil, err
+	}
+	return &StoreLock{Key: key, owner: owner, lockedWithKey: true}, nil
+}
+
+func (k8s *k8sStore) Unlock(storeLock *StoreLock) error {
+	if storeLock.lockedWithKey {
+		return k8s.cm.UnlockWithKey(storeLock.Key)
+	}
+	return k8s.cm.Unlock()
+}
+
+func (k8s *k8sStore) IsKeyLocked(key string) (bool, string, error) {
+	return k8s.cm.IsKeyLocked(key)
+}
+
+func (k8s *k8sStore) CreateKey(key string, value []byte) error {
+	sanitizedKey := GetSanitizedK8sName(key)
```

Comment:
GetSanitizedK8sName was added as part of a different request.
The only time it will result in same sanitized key for two different keys is when the two keys differ in a special character.
Key@ and Key$.
At the time of implementation we determined that special characters shouldn't be a differentiator. The idea of mapping each special character to some Letter was also pitched but wasn't perceived to be necessary.